File: solutions/clojure/013-3sum.clj
;; Problem 13: 3Sum (LeetCode 15)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn three-sum [nums]
  (let [sorted (vec (sort nums))
        n (count sorted)]
    (loop [i 0, result []]
      (if (>= i n) result
        (if (and (> i 0) (= (sorted i) (sorted (dec i))))
          (recur (inc i) result)
          (let [target (- (sorted i))
                trips (loop [l (inc i), r (dec n), acc []]
                        (if (>= l r) acc
                          (let [s (+ (sorted l) (sorted r))]
                            (cond
                              (< s target) (recur (inc l) r acc)
                              (> s target) (recur l (dec r) acc)
                              :else (let [nl (loop [ll (inc l)]
                                            (if (and (< ll r) (= (sorted ll) (sorted l)))
                                              (recur (inc ll)) ll))]
                                     (recur nl (dec r)
                                       (conj acc [(- target) (sorted l) (sorted r)])))))))]
            (recur (inc i) (into result trips))))))))

File: solutions/rust/013-3sum.rs
/**
 * Problem 13: 3Sum (LeetCode 15)
 * Difficulty: Med
 * Language: Rust
 */
pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
    nums.sort();
    let n = nums.len();
    let mut result = Vec::new();
    for i in 0..n {
        if i > 0 && nums[i] == nums[i - 1] { continue; }
        let (mut l, mut r) = (i + 1, n.wrapping_sub(1));
        while l < r {
            let sum = nums[i] + nums[l] + nums[r];
            match sum.cmp(&0) {
                std::cmp::Ordering::Less => l += 1,
                std::cmp::Ordering::Greater => r -= 1,
                std::cmp::Ordering::Equal => {
                    result.push(vec![nums[i], nums[l], nums[r]]);
                    while l < r && nums[l] == nums[l + 1] { l += 1; }
                    l += 1; r -= 1;
                }
            }
        }
    }
    result
}
